hal/api/toolchain.h

```diff
@@ -16,6 +16,214 @@
 #ifndef MBED_TOOLCHAIN_H
 #define MBED_TOOLCHAIN_H
 
+
+// Warning for unsupported compilers
+#if !defined(__GNUC__)   /* GCC        */ \
+ && !defined(__CC_ARM)   /* ARMCC      */ \
+ && !defined(__clang__)  /* LLVM/Clang */ \
+ && !defined(__ICCARM__) /* IAR        */
+#warning "This compiler is not yet supported."
+#endif
+
+
+// Attributes
+
+/** MBED_PACKED
+ *  Pack a structure, preventing any padding from being added between fields.
+ *
+ *  @code
+ *  #include "toolchain.h"
+ *
+ *  typedef struct {
+ *      char x;
+ *      int y;
+ *  } MBED_PACKED foo;
+ *  @endcode
+ */
+#ifndef MBED_PACKED
+#if defined(__ICCARM__)
+#define MBED_PACKED __packed
```

Comment:
I remember some compilers (might not be in our code base) require packed attribute to be in the beginning or end, therefore modules like lwip define packed as ``PACK_STRUCT_BEGIN`` , ``PACK_STRUCT_STRUCT`` and ``PACK_STRUCT_END``.